Use ToggleEvent for <details> element toggle event

sideshowbarker opened the pull request:
#### 74637c383d415af9b4e4ca2ce7fbb3268e6ead85
<pre>
Use ToggleEvent for &lt;details&gt; element toggle event
<a href="https://bugs.webkit.org/show_bug.cgi?id=252373">https://bugs.webkit.org/show_bug.cgi?id=252373</a>

Reviewed by Tim Nguyen.

This change makes ToggleEvent be used for the event that’s fired when a
“details” element is opened or closed, while handling the case of
coalesced events.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/resources/import-expectations.json:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/toggleEvent-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/toggleEvent.html:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/w3c-import.log:
* Source/WebCore/html/HTMLDetailsElement.cpp:
(WebCore::HTMLDetailsElement::queueDetailsToggleEventTask):
(WebCore::HTMLDetailsElement::attributeChanged):
* Source/WebCore/html/HTMLDetailsElement.h:

Canonical link: <a href="https://commits.webkit.org/267076@main">https://commits.webkit.org/267076@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/25324c12edc811eb5c04f9991bf5ac790231b489

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/15585 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/15891 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/16256 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/17341 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/14620 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/15767 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/18833 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/15987 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/17173 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/15770 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/16225 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/13257 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/18086 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/13490 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/14059 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/20988 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/14515 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/14226 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/17494 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/14816 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/12554 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/14072 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/13906 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3733 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/18434 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/14636 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/w3c-import.log:
```diff
@@ -23,6 +23,7 @@ List of files:
 /LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/details-add-summary.html
 /LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/details-cq-crash.html
 /LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/details-findstring-crash.html
+/LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/details-keyboard-activation.html
```
This file doesn't seem to be in the patch.

nt1m commented on Source/WebCore/html/HTMLDetailsElement.cpp:
```diff
@@ -130,6 +131,27 @@ bool HTMLDetailsElement::isActiveSummary(const HTMLSummaryElement& summary) cons
+void HTMLDetailsElement::queueDetailsToggleEventTask(DetailsState oldState, DetailsState newState)
+{
+    if (m_isToggleEventTaskQueued)
+        return;
```
I think `m_isToggleEventTaskQueued` should be entirely removed in order for event coalescing to work correctly. Otherwise, you'll always end up just emitting the first event, instead of an event containing the first old state and the last new state.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/details-keyboard-activation.html:
Test results for this test need to be generated 
```
run-webkit-tests imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/details-keyboard-activation.html

git add imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/details-keyboard-activation-expected.txt
```

nt1m commented on Source/WebCore/html/HTMLDetailsElement.h:
```diff
@@ -52,6 +68,9 @@ class HTMLDetailsElement final : public HTMLElement {
     WeakPtr<HTMLSummaryElement, WeakPtrImplWithEventTargetData> m_defaultSummary;
     RefPtr<HTMLSlotElement> m_defaultSlot;
     bool m_isToggleEventTaskQueued { false };
```
This can be removed.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/toggleEvent.html:
```diff
@@ -66,7 +66,7 @@
     assert_false(evt.bubbles, "event doesn't bubble");
     assert_false(evt.cancelable, "event is not cancelable");
-    assert_equals(Object.getPrototypeOf(evt), Event.prototype, "Prototype of toggle event is Event.prototype");
+    assert_equals(Object.getPrototypeOf(evt), ToggleEvent.prototype, "Prototype of toggle event is ToggleEvent.prototype");
```
Can we assert the oldState / newState property is set correctly every time as well?

Also, this will need to be exported to WPT (export-w3c-test-changes can help you with that)

sideshowbarker replied:
> Can we assert the oldState / newState property is set correctly every time as well?

Yep, I will add those

> Also, this will need to be exported to WPT (export-w3c-test-changes can help you with that)

aha, cool — will do (after review approval of the test updates here)

nt1m replied:
I don't see the test updates here yet

sideshowbarker replied:
> I don't see the test updates here yet

Yeah, sorry, I’m still working on the test updates. (The most-recent force push was just to update/amend the commit message, not any code.)

nt1m replied:
The newer revision still doesn't seem to have the assertions for old/newState

sideshowbarker replied:
> The newer revision still doesn't seem to have the assertions for old/newState

Right — that’s because when I added those assertions to the tests locally and ran them, the old/newState assertions I added to the two tests for _removing_ the `open` attribute are failing; specifically, my local WebKit build seems unexpectedly seems to be causing both the `newState` _and_ the `oldState` to be set to `closed` when the `open` attribute is removed.

So I guess I need to do more troubleshooting locally before I push the updated tests to this PR branch (because otherwise they’re just going to cause all the EWS bots to go red…)

sideshowbarker replied:
OK, in https://github.com/WebKit/WebKit/pull/16795/commits/5eb916675606f6af3bbbf5a981531ceaa8ab8729 I added the expected old/newState to all the _“adding open”_ tests.

However, I didn’t yet do the same for the _“removing open”_ tests, nor for the _“Calling open twice on 'details' fires only one toggle event”_ — because for those tests I’m not getting the expected states. Details:

- _“Removing open from 'details' should fire a toggle event at the 'details' element”_ (details2/t2)
  - expected: `oldState: open, newState: closed`
  - actual: `oldState: closed, newState: closed` ❌
- _“Calling removeAttribute('open') to 'details' should fire a toggle event at the 'details' element”_ (details8/t8)
  - expected: `oldState: open, newState: closed`
  - actual: `oldState: closed, newState: closed` ❌
  
_“Calling open twice on 'details' fires only one toggle event”_ also isn’t giving the expected results:
  - expected: `oldState: closed, newState: open`
  - actual: `oldState: closed, newState: closed` ❌

I’ll do some debugging, and post details from that in another comment.

sideshowbarker replied:
OK, in https://github.com/WebKit/WebKit/commit/84714b94de65e8425bea8c09b7091ebaa862d9b0 I added old/newState assertions for all tests, and fixed the code to produce the expected results.

nt1m commented on Source/WebCore/html/HTMLDetailsElement.cpp:
```diff
@@ -138,20 +156,13 @@ void HTMLDetailsElement::attributeChanged(const QualifiedName& name, const AtomS
+            if (m_isOpen) {
                 root->appendChild(*m_defaultSlot);
-            else
+                queueDetailsToggleEventTask(DetailsState::Open, DetailsState::Closed);
```
This looks backwards, oldState (first param) should be closed, and newState should be open

nt1m commented on Source/WebCore/html/HTMLDetailsElement.cpp:
```diff
@@ -138,20 +156,13 @@ void HTMLDetailsElement::attributeChanged(const QualifiedName& name, const AtomS
-                m_isToggleEventTaskQueued = false;
-            });
-            m_isToggleEventTaskQueued = true;
+                queueDetailsToggleEventTask(DetailsState::Closed, DetailsState::Open);
```
ditto, this also looks backwards

nt1m reviewed (approved):
Can we export the WPT changes so they don't get lost?

Looks good otherwise.

sideshowbarker committed:
Use ToggleEvent for <details> element toggle event
https://bugs.webkit.org/show_bug.cgi?id=252373

Reviewed by Tim Nguyen.

This change makes ToggleEvent be used for the event that’s fired when a
“details” element is opened or closed, while handling the case of
coalesced events.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/resources/import-expectations.json:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/toggleEvent-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/toggleEvent.html:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/w3c-import.log:
* Source/WebCore/html/HTMLDetailsElement.cpp:
(WebCore::HTMLDetailsElement::queueDetailsToggleEventTask):
(WebCore::HTMLDetailsElement::attributeChanged):
* Source/WebCore/html/HTMLDetailsElement.h:

Canonical link: https://commits.webkit.org/267076@main
